build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -1288,6 +1315,218 @@ message WaitExecutionRequest {
+      // this pathset does not match the filesystem. Otherwise the hash of the string constant "E"
+      // followed by a hash of the directory contents filtered to only the file names in
+      // ObservedAccessFileNames (compared to only the directory's filenames without extensions)
+      // must be added to the prospective strong fingerprint. See comments on
+      // [ObservedPathSet][build.bazel.remote.execution.v2.ObservedPathSet] for how to generate a
+      // directory hash.
+      DIRECTORY_SEARCH = 2;
+
+      // 'Path' is a file probed for existence, but not read, by the action and that was present
+      // during execution. If the file is not present or is instead a directory, this pathset
+      // does not match the filesystem. Otherwise the hash of the string constant "F" must be
+      // added to the prospective strong fingerprint.
+      EXISTING_FILE_PROBE = 3;
+
+      // `Path` is a directory probed for existence, but not enumerated, by the action and that
+      // was present during execution. If the directory is not present, or is instead a file,
+      // this pathset does not match the filesystem. Otherwise the hash of the string constant "D"
+      // must be added to the prospective strong fingerprint.
+      EXISTING_DIRECTORY_PROBE = 4;
+
+      // `Path` is a file or directory probed for existence by the action that was not
+      // present during execution. If path is present as a file or directory, this pathset
+      // does not match the filesystem. Otherwise the hash of the string constant "P" must
+      // be added to the prospective strong fingerprint.
+      ABSENT_FILE_OR_DIRECTORY_PROBE = 5;
+    }
+
+    // File or directory path.
+    string Path = 1;
+
+    // Defines the meaning of `Path` and how to compare to the local filesystem.
+    OpKind Kind = 2;
+
+    // For entries of `OpType.DIRECTORY_ENUMERATION`, the enumeration pattern used.
+    // When null/empty/missing the search pattern was "*" meaning the full directory.
+    string DirectorySearchPattern = 3;
+  }
+
+  // Observed path set entries. In order to ensure consistent hashing of the same `ObservedPathSet`,
+  // the entries MUST be sorted lexicographically by Path code point (or, equivalently, by UTF-8
+  // bytes).
+  repeated Entry Entries = 1;
+
+  // For directory comparisons in Entries marked with `OpKind.DIRECTORY_SEARCH`, defines the filenames,
+  // with extensions removed, that were read by the action. These are used as the filter for generating
+  // a directory hash to add to the prospective strong fingerprint.
+  repeated string ObservedAccessedFileNames = 2;
+}
+
+TODO: erik: Add AnyBuild 2-way exec contract and discuss VFS layouts
+
+// Represents a two-phase cache pathset, referenced or inlined, with the strong fingerprint
+// that can be used to retrieve build outputs if the pathset matches the client's filesystem.
+message ObservedPathSetAndStrongFingerprint {
```

Comment:
>ObservedPathSetAndStrongFingerprint [](start = 8, length = 35)

Add expiry.

Reply:
If we do want to cache info locally, we know how long we're allowed to keep that information allowed before re-query against the datacenter. NOTE: Perhaps this belongs in Capabilities?

---
In reply to: [453829743](https://github.com/erikma/remote-apis/pull/1#discussion_r453829743) [](ancestors = 453829743)

Reply:
Lance - can you paste what the typical timeout for this looks like in terms of time? Alternately do you think this is something we can leave to the implementation? E.g. a client implementation could have its own implementation-specific timeout instead of having the agent/service side try to control things.

---
In reply to: [453830620](https://github.com/erikma/remote-apis/pull/1#discussion_r453830620) [](ancestors = 453830620,453829743)

Reply:
Thinking about this, I think this is an implementation detail we can leave to client configuration, and not add either to the per-action protocol or the capabilities API as a proposed extension. Please reopen the conversation if you disagree.